web/chrome_print_service_extension.js

```diff
@@ -0,0 +1,477 @@
+   * 9. CS: Use X.postMessage to send [nonceForAuthX, port2] to X, WITH
+   *          extension origin as origin target.
+   *
+   *       Steps 10+13 and step 11+12 run parallelly.
+   *
+   * 10.F: Receive message, proceed iff nonceForAuthC' == nonceForAuthC.
+   * 11.F: Use port1.postMessage to send [nonceForAuthF] to X.
+   *
+   * ------------- Now F knows that port1 is from CS -------------------------
+   *       Note: port1 was directly sent to an extension origin, so no further
+   *       authentication is needed on port1.
+   *
+   * 12.X: Receive message, proceed iff nonceForAuthX' == nonceForAuthX and
+   *       if message came from the top frame (CS).
+   *
+   * ------------- Now X knows that port2 is from CS -------------------------
+   *
+   * 13.X: Receive on port2, proceed iff nonceForAuthF' == nonceForAuthF
+   *
+   * ------------- Now X knows that port2 is entangled with F ----------------
+   * ------------- We have a secure channel between X and F ------------------
+   *
+   * 14.X: Send print job including Blobs over the MessageChannel to F.
+   * 15.F: Receive print job.
+   * 16.F: Create the document referencing these blobs and start printing.
+   *
+   * At any point either X or the frame can send a message with nonceToAbort
+   * to signal that the print request should be canceled.
+   */
+
+  /**
+   * Alternatives
+   *
+   * The following alternatives have been considered and rejected:
+   * - Any combination where the origin of the frame equals the main frame,
+   *   because of information leakage.
+   * - Creating blob:-URLs in the main frame, navigate frame to data: or blob:
+   *   URL referencing the images at blob:. Does not work, cross-origin blob:
+   *   URLs won't load.
+   * - Serializing images as data:-URLs, create one giant data:-URL and
+   *   navigate the document. Max a few dozen lines of code, but using lots of
+   *   memory to hold the string.
+   * - Open PDF in new tab, and print from there. Bad UX, no thanks.
+   * - Navigate to my website, print from there. No, the extension should be
+   *   self-contained. And if --site-per-process is enabled then the OOP
+   *   printing problem is back.
+   */
+
+  // This is only for debugging. The function is also serialized and passed to
+  // the content script (CS), which also passes it to frame F.
+  function LogStep(step, description) {
+    console.info('Step ' + step + '. ' + description);
+  }
+
+  // This function is serialized and run in the main frame in step 2.
+  function topContentScript(nonceNoDupe, nonceForAuthX, nonceForAuthC,
+    nonceForAuthF, nonceToAbort) {
+    var EXTENSION_ORIGIN = 'chrome-extension://' + chrome.runtime.id;
+
+    var printFrame = null;
+
+    window.addEventListener('message', onmessage, true);
+    function onmessage(event) {
+      if (event.origin === EXTENSION_ORIGIN && event.data === nonceNoDupe) {
+        LogStep(4, 'Receive message from X');
```

Comment:
Travis complains about these, you probably need to add `LogStep` to the list of `globals`.